### ISSUE TITLE
Experiments replication

### ISSUE THREAD

**Issue by gxenos** [chamathpali/FedSim#4]
Hi, 

I am trying to replicate the experiments from the original paper on the FEMINST dataset using multinomial logistic regression (I am implementing your method as a baseline using the flower framework for FL). I was wondering what are the dataset parameters you used in the experiment? In the paper you mention "A subset of FEMNIST representing 26 classes is used in the FL setting as proposed by [14]. A total of 18,345 samples are distributed among 200 clients and each client has samples only for 5 classes."

But in your code you have 10 classes (characters 'a' -'j') and each client has samples only for 3 classes.

Can you help me understand what are the correct values?

Thanks

**Comment by chamathpali** [comment]
Hi @gxenos, apologies for the confusion. This was an writing error, The paragraph was trying to describe the original dataset in detail, original dataset size is `18,345` (before pre-processing) and FedProx [14] was using 5 classes. In our experiments we had created multiple variants of the datasets for further analysis and the final class size used is as in the code **3**. 
The main reason for trying different class size variants is, as discussed in the paper FedSim performs best when there is moderate level of statistical heterogeneity

Also, here is an updated snippet from our new work using the same dataset,

> FEMNIST (Federated-Extended-MNIST) is a handwritten character recognition dataset from [ ]. The subsample consists of 10 lowercase characters (a-j) and is used for a 10-class character classification task. This dataset is distributed among 200 clients, with each client having samples for only 3 classes. Each data instance in this dataset is an image with dimensions of 28x28

In our latest work we have fixed the typo and added further experiments on class sizes 1,2,3  (Section 6.6) 
https://arxiv.org/pdf/2409.05242 if you are interested.

Let me know if this clears up your question and if you want any support on implementing the baseline..

**Comment by gxenos** [comment]
@chamathpali thanks for the quick reply. That makes sense. 

I am trying to recreate the results in this diagram from the FedSim paper.  

<img width="631" height="474" alt="Image" src="https://github.com/user-attachments/assets/47eaa6b0-b0d0-4af0-9c30-711877b6980b" />

Based on your code and this line:
`num_samples = np.random.lognormal(4, 1, (NUM_USER)) + 5`

 I have implemented a dataset with 10 lowercase characters (a-j), 3 classes per client and with the number of samples per client as described above. That leads to a dataset with 18,345 total samples distributed to 200 clients. 

For each round 20 clients are selected for participation, forming 9 clusters, and after each round test accuracy is measured on all 200 clients and averaged (weighted by the number of samples). 

The results I get however are not that smooth and are more similar to this:

<img width="453" height="463" alt="Image" src="https://github.com/user-attachments/assets/f622702e-8985-4f77-adc7-ae00173705f5" />

which are reported in the baseline implementation of FedProx in the Flower repo and are similar to the results reported in the FedProx paper. 

Are you averaging the results per round over multiple experiments? 

> The experiments performed on all the datasets were carried out with 35 random seeds (from 0 to 34 incremented by 1) 

Or am I missing something else? 

In any case FedSim seems to outperform FedAvg, similar to the original first diagram.

**Comment by chamathpali** [comment]
@gxenos here is a sample of a single run just for demo purposes https://github.com/chamathpali/FedSim/blob/main/logs/sample/nist_0_fedsim.pdf
And yes the graphs are averaged out for 35 runs for significance tests.

Also here is the code used to plot the graphs going over all the runs with diff random seeds
https://github.com/chamathpali/FedSim/blob/main/plot_fedsim_main.py

**Comment by gxenos** [comment]
Ok great! This is very similar to what I get. Thanks! Will let you know when I have the final results.

**Comment by gxenos** [comment]
@chamathpali So this is what I get with Flower. 

<img width="572" height="426" alt="Image" src="https://github.com/user-attachments/assets/43461f4f-3d62-4fbc-a0b5-4dd7cb070abd" />

While FedSim has superior performance, my results are slightly different. I did average the accuracies for each round but still the plot is not that smooth. The same applies for FedAvg where I am using the official Flower implementation. 

Is my experimental setup different or is it due to the framework?

**My experimental setup is the following:**

- 200 participants - 20 selected each round
-  Dataset: FEMNIST subsampled to 10 lowercase characters (a-j)
- Each participant holds samples for only 3 classes

This setup leads to 18,345 samples being distributed among 200 clients.

- Model: MLR
`self.linear = nn.Linear(28*28, 10)`
- lr=0.003
- local_epochs=20
- rounds=500

I measure the accuracy of the produced model, each round, on all (200) clients' private test data and average (weighted with the number of samples). 

- Train-Test split 80-20

I was expecting that at least for FedAvg I would get the exact same results (or something very similar).

**Comment by chamathpali** [comment]
@gxenos this is for a single run right? the smooth graphs are when run for 35 runs with different random seeds and averaged

**Comment by gxenos** [comment]
Unfortunately this is the averaged for 35 seeds.

**Comment by chamathpali** [comment]
can you show two graphs from two individual runs to get a better idea?

**Comment by gxenos** [comment]
Sure. This is for the same seed.

<img width="561" height="423" alt="Image" src="https://github.com/user-attachments/assets/5f1be29d-5be0-47e1-95f7-bbcd54764ec3" />

**Comment by chamathpali** [comment]
@gxenos do you have any open source code for the experiments to check it out?

**Comment by gxenos** [comment]
@chamathpali thanks for the reply! 

Sure, the code is here: https://github.com/gxenos/flower/tree/add-fedsim-baseline/baselines/fedsim/fedsim

This file contains the dataset partitioning logic: https://github.com/gxenos/flower/blob/add-fedsim-baseline/baselines/fedsim/fedsim/dataset.py

While this one defines the client behavior: https://github.com/gxenos/flower/blob/add-fedsim-baseline/baselines/fedsim/fedsim/client_app.py

The implementation of the FedSim strategy is here: https://github.com/gxenos/flower/blob/add-fedsim-baseline/baselines/fedsim/fedsim/strategy.py, but I believe the problem lies in the above-mentioned files since I get a similar non-smooth result even when using the default FedAvg provided by the framework
